cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_dense_default_impl.i

```diff
@@ -61,39 +61,70 @@ using namespace kdtree_knn_classification::internal;
 template <typename T, CpuType cpu>
 class Queue
 {
+    static const size_t defaultSize = 4;
+
 public:
-    Queue() : _data(nullptr) {}
+    Queue() : _data(nullptr), _first(0), _last(0), _count(0), _size(0), _capacity(0) {}
 
-    ~Queue()
-    {
-        services::daal_free(_data);
-        _data = nullptr;
-    }
+    ~Queue() { clear(); }
+
+    Queue(const Queue &)             = delete;
+    Queue & operator=(const Queue &) = delete;
 
     bool init(size_t size)
     {
         clear();
+        if (size == 0) // Check for valid size
+        {
+            return false;
+        }
+
         _first = _count = 0;
         _last = _sizeMinus1 = (_size = size) - 1;
-        return ((_data = static_cast<T *>(service_malloc<T, cpu>(size * sizeof(T)))) != nullptr);
+        _data               = static_cast<T *>(service_malloc<T, cpu>(size));
+
+        if (!_data) // Check if memory allocation was successful
+        {
+            return false;
+        }
+
+        _capacity = _size; // Initialize capacity
+        return true;
     }
 
     void clear()
     {
-        daal_free(_data);
-        _data = nullptr;
+        if (_data)
+        {
+            daal::services::internal::service_free<T, cpu>(_data); // Free allocated memory if it exists
+            _data = nullptr;
+        }
+        _first = _last = _count = _size = _sizeMinus1 = _capacity = 0; // Reset state
     }
 
+    void reset() { _first = _last = _count = 0; }
+
     DAAL_FORCEINLINE void push(const T & value)
     {
-        _data[_last = (_last + 1) & _sizeMinus1] = value;
+        if (_count >= _capacity) // Check if capacity is exceeded
+        {
+            services::Status status = grow(); // Grow if necessary
+            //DAAL_CHECK_STATUS_VAR(status);
+        }
+
+        _data[_last = (_last + 1) & _sizeMinus1] = value; // Add element to queue
```

Comment:
It seems that for this condition to make sense Size should be the power of 2. Since this can depend on value provided by user it is better to use (_last + 1) % size

Reply:
Or explicitly force size to be the power of two

Reply:
the queue class has been updated